sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -767,6 +777,35 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
   override protected def withNewChildInternal(newChild: Expression): InSet = copy(child = newChild)
 }
 
+object InSet {
+  class LCaseSet(inputSet: Set[Any]) extends immutable.Set[Any] with Serializable {
+    private val strSet = inputSet.map { s =>
+      if (s == null) null
+      else CollationAwareUTF8String.lowerCaseCodePoints(s.asInstanceOf[UTF8String])
+    }
+    override def incl(elem: Any): Set[Any] = inputSet.incl(elem)
+    override def excl(elem: Any): Set[Any] = inputSet.excl(elem)
+    override def iterator: Iterator[Any] = inputSet.iterator
+    override def contains(elem: Any): Boolean = {
+      assert(elem != null, "InSet guarantees non-null input")
+      strSet.contains(CollationAwareUTF8String.lowerCaseCodePoints(elem.asInstanceOf[UTF8String]))
+    }
+  }
+  class CollationSet(inputSet: Set[Any], collationId: Int)
+    extends immutable.Set[Any] with Serializable {
+    override def incl(elem: Any): Set[Any] = inputSet.incl(elem)
+    override def excl(elem: Any): Set[Any] = inputSet.excl(elem)
+    override def iterator: Iterator[Any] = inputSet.iterator
+    override def contains(elem: Any): Boolean = {
+      assert(elem != null, "InSet guarantees non-null input")
+      val collation = CollationFactory.fetchCollation(collationId)
+      inputSet.exists { p =>
```

Comment:
This seems pretty inefficient, can we also do the similar thing as we did for the lcase one -> preprocess all elements by calculating the hashCode of the string's collationKey and then create a Map[Int, Seq[String]].

After that, in contains we can query the map to see if the element is present or not.

Reply:
cc: @uros-db who did some work with collation keys to see if this makes sense

Reply:
yeah, for lots of comparisons we should use collation key
https://unicode-org.github.io/icu/userguide/collation/concepts#sortkeys-vs-comparison

Reply:
> This seems pretty inefficient

It depends on the cost of key calculation, size of the input set, how much are comparisons of the particular collation heavy. This could seems like "premature optimizations", but maybe you are right.

Reply:
I changed implementation, but I think the old implementation can overperform new one in some cases. Need to benchmark both.